### PR TITLE
EVM RPC

### PR DESCRIPTION
**Pull request by ermalkaleci**
Add support for `eth_estimateGas` and `eth_call`

### CODE REVIEW

File: rpc/evm/runtime_api/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+#![allow(clippy::all)]
+use ethereum_types::{H160, U256};
+use sp_std::vec::Vec;
+
+sp_api::decl_runtime_apis! {
+	pub trait EVMApi {
+		fn call(
```
Comment (xlc) [inline rpc/evm/runtime_api/src/lib.rs:8]: unify call & create & create2 into one

File: rpc/evm/runtime_api/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+			from: H160,
+			to: H160,
+			data: Vec<u8>,
+			value: U256,
+			gas_limit: U256,
+			gas_price: U256,
```
Comment (xlc) [inline rpc/evm/runtime_api/src/lib.rs:14]: remove this

File: rpc/evm/runtime_api/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+			nonce: Option<U256>,
```
Comment (xlc) [inline rpc/evm/runtime_api/src/lib.rs:15]: remove this

File: rpc/evm/runtime_api/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+use sp_std::vec::Vec;
+
+sp_api::decl_runtime_apis! {
+	pub trait EVMApi {
+		fn call(
+			from: H160,
+			to: H160,
+			data: Vec<u8>,
+			value: U256,
+			gas_limit: U256,
```
Comment (xlc) [inline rpc/evm/runtime_api/src/lib.rs:13]: change type to u64

File: rpc/evm/src/bytes.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
+//! Serializable wrapper around vector of bytes
```
Comment (xlc) [inline rpc/evm/src/bytes.rs:1]: we shouldn't need this file

File: rpc/evm/src/call_request.rs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#[derive(Debug, Default, PartialEq, Deserialize)]
+#[serde(deny_unknown_fields)]
+#[serde(rename_all = "camelCase")]
+pub struct CallRequest {
+	/// From
+	pub from: Option<H160>,
+	/// To
+	pub to: Option<H160>,
+	/// Gas Price
+	pub gas_price: Option<U256>,
```
Comment (xlc) [inline rpc/evm/src/call_request.rs:15]: remove this

File: rpc/evm/src/call_request.rs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+	/// Gas
+	pub gas: Option<U256>,
+	/// Value
+	pub value: Option<U256>,
+	/// Data
+	pub data: Option<Bytes>,
+	/// Nonce
+	pub nonce: Option<U256>,
```
Comment (xlc) [inline rpc/evm/src/call_request.rs:23]: remove this

File: rpc/evm/src/call_request.rs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+#[serde(rename_all = "camelCase")]
+pub struct CallRequest {
+	/// From
+	pub from: Option<H160>,
+	/// To
+	pub to: Option<H160>,
+	/// Gas Price
+	pub gas_price: Option<U256>,
+	/// Gas
+	pub gas: Option<U256>,
```
Comment (xlc) [inline rpc/evm/src/call_request.rs:17]: change to u64

File: rpc/evm/src/evm_api.rs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+use jsonrpc_core::Result;
+use jsonrpc_derive::rpc;
+
+pub use rpc_impl_EVMApi::gen_server::EVMApi as EVMApiServer;
+
+/// EVM rpc interface.
+#[rpc(server)]
+pub trait EVMApi<BlockHash> {
+	/// Call contract, returning the output data.
+	#[rpc(name = "eth_call")]
```
Comment (xlc) [inline rpc/evm/src/evm_api.rs:14]:
```suggestion
	#[rpc(name = "evm_call")]
```

merge both into one and change return type to `Result<(Bytes, u64)>`

**Review by xlc — APPROVED** [review]

**Comment by xlc** [comment]
@ermalkaleci you can do all the commented changes in a new PR

**Comment by ermalkaleci** [comment]
@xlc yes I have another draft request addressing those, branch module-evm

**Pull request merged**